docs/reference/how-to/size-your-shards.asciidoc

```diff
@@ -175,17 +175,23 @@ index                                 prirep shard store
 
 [discrete]
 [[shard-count-recommendation]]
-==== Aim for 20 shards or fewer per GB of heap memory
+==== Aim for 3000 indices or fewer per GB of heap memory on master nodes
 
-The number of shards a data node can hold is proportional to the node's heap
-memory. For example, a node with 30GB of heap memory should have at most 600
-shards. The further below this limit you can keep your nodes, the better. If
-you find your nodes exceeding more than 20 shards per GB, consider adding
-another node.
+The number of indices a master node can manage is proportional to its heap
+size. The exact amount of heap memory needed for each index depends on various
+factors such as the size of the mapping and the number of shards per index.
 
-Some system indices for {enterprise-search-ref}/index.html[Enterprise Search]
-are nearly empty and rarely used. Due to their low overhead, you shouldn't
-count shards for these indices toward a node's shard limit.
+Where possible, the master node will deduplicate the mapping metadata across
+indices with identical mappings. Mapping metadata will be deduplicated across
+indices which are created from the same index template and
+<<explicit-mapping,do not use dynamic mapping updates>>.
+
+If your indices are are mostly created from a small number of templates and do
```

Comment:
I think of something pretty simple and conservative. Perhaps just 1-2kb per field?

For dynamic mappings, it can make the impact of those very visible. Like you need a 1GB heap if you do not use them but a 30GB heap (or whatever the calculation says) if you use them. That can be a good motivation for a redesign?

Another case could be applications or search use cases using 100s of specific indices with specific mappings. They fall outside the above and we then leave them with no guidance. I think those cases will definitely also result in pretty small numbers anyway, but providing the guidance seems helpful.

Reply:
I think it might make sense to defer this until https://github.com/elastic/elasticsearch/issues/86639 is addressed - IMO it's too complicated to explain here how dynamic mappings might work with the deduplication mechanism.

Reply:
I would set it up as a worst-case with no deduplication. I agree that anything in between is hard, but providing a no-sharing guidance should be relatively easy? It puts an upper boundary on the no-sharing required heap-size.

Reply:
> but providing a no-sharing guidance should be relatively easy? It puts an upper boundary on the no-sharing required heap-size.

But this totally depends on the size of the mappings used. Also, the effect of not sharing isn't as large as it is made out to be here I think. A compressed beats mapping for example is only *~6k*. So for 3k indices we're only talking about an 18M outright increase in the direct heap requirements for the master node. There's additional effects here from making stats APIs heavier without dedup, sending larger cluster states etc. but I find it very hard to capture these in a heap size number.

Reply:
We talked offline about this and it seems we could remove the deduplication assumption here, since it is in effect very little extra data on the masters and is more of a cpu concern than memory.